hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenFreezeAccountHandler.java

```diff
@@ -44,23 +54,99 @@ public TokenFreezeAccountHandler() {
         final var tokenMeta = tokenStore.getTokenMeta(op.tokenOrElse(TokenID.DEFAULT));
         if (tokenMeta == null) throw new PreCheckException(INVALID_TOKEN_ID);
         if (tokenMeta.hasFreezeKey()) {
             context.requireKey(tokenMeta.freezeKey());
+        } else {
+            throw new PreCheckException(TOKEN_HAS_NO_FREEZE_KEY);
         }
     }
 
     /**
      * This method is called during the handle workflow. It executes the actual transaction.
      *
-     * <p>Please note: the method signature is just a placeholder which is most likely going to
-     * change.
-     *
+     * @param txn the {@link TokenFreezeAccountTransactionBody} of the active transaction
+     * @param accountStore the {@link ReadableAccountStore} for the active transaction
+     * @param tokenStore the {@link ReadableTokenStore} for the active transaction
+     * @param tokenRelStore the {@link WritableTokenRelationStore} for the active transaction
      * @throws NullPointerException if one of the arguments is {@code null}
      */
-    public void handle() {
-        throw new UnsupportedOperationException("Not implemented");
+    public void handle(
+            @NonNull final TransactionBody txn,
+            @NonNull ReadableAccountStore accountStore,
+            @NonNull ReadableTokenStore tokenStore,
+            @NonNull WritableTokenRelationStore tokenRelStore)
+            throws HandleException {
+        requireNonNull(txn);
+        requireNonNull(accountStore);
+        requireNonNull(tokenStore);
+        requireNonNull(tokenRelStore);
+
+        final var op = txn.tokenFreezeOrThrow();
+        final var tokenRel = validateSemantics(op, accountStore, tokenStore, tokenRelStore);
+
+        final var copyBuilder = tokenRel.copyBuilder();
+        copyBuilder.frozen(true);
+        tokenRelStore.put(copyBuilder.build());
+    }
+
+    /**
+     * Performs checks independent of state or context
+     */
+    private void pureChecks(TokenFreezeAccountTransactionBody op) throws PreCheckException {
+        if (!op.hasToken()) {
+            throw new PreCheckException(INVALID_TOKEN_ID);
+        }
+
+        if (!op.hasAccount()) {
+            throw new PreCheckException(INVALID_ACCOUNT_ID);
+        }
+    }
+
+    /**
+     * Performs checks that the given token and accounts from the state are valid and that the
+     * token is associated to the account
+     *
+     * @return the token relation for the given token and account
+     */
+    private TokenRelation validateSemantics(
+            TokenFreezeAccountTransactionBody op,
+            ReadableAccountStore accountStore,
+            ReadableTokenStore tokenStore,
+            WritableTokenRelationStore tokenRelStore)
+            throws HandleException {
+        // Check that the token exists
+        final var tokenId = op.tokenOrElse(TokenID.DEFAULT);
+        final ReadableTokenStore.TokenMetadata token;
+        try {
+            token = tokenStore.getTokenMeta(tokenId);
```

Comment:
I don't think we throw `PrecheckException` from `getTokenMeta`  🤔 . We can remove throw there and remove try-catch here

Reply:
Doesn't compile otherwise...
<img width="623" alt="Screenshot 2023-05-05 at 09 20 44" src="https://user-images.githubusercontent.com/112430600/236499804-5641377c-7803-4f94-9efc-8ab351afbc6b.png">

Reply:
Removed `throws PreCheckException` from the token store as discussed